python/ray/tune/trial_runner.py

```diff
@@ -109,45 +113,46 @@ def _launch_trial(self):
         trial = self._get_runnable()
         self._commit_resources(trial.resources)
         try:
-            trial.start()
-            self._pending[trial.train_remote()] = trial
+            trial.start(path=self._trial_restore_paths[trial])
+            self._running[trial.train_remote()] = trial
         except:
             print("Error starting agent, retrying:", traceback.format_exc())
             time.sleep(2)
             trial.stop(error=True)
             try:
                 trial.start()
-                self._pending[trial.train_remote()] = trial
+                self._running[trial.train_remote()] = trial
             except:
                 print("Error starting agent, abort:", traceback.format_exc())
                 trial.stop(error=True)
                 # note that we don't return the resources, since they may
                 # have been lost
 
     def _process_events(self):
-        [result_id], _ = ray.wait(self._pending.keys())
-        trial = self._pending[result_id]
-        del self._pending[result_id]
+        [result_id], _ = ray.wait(self._running.keys())
+        trial = self._running[result_id]
+        del self._running[result_id]
         try:
             result = ray.get(result_id)
             print("result", result)
             trial.last_result = result
 
             if trial.should_stop(result):
-                self._return_resources(trial.resources)
-                trial.stop()
+                self._stop_trial(trial)
             else:
                 # TODO(rliaw): This implements checkpoint in a blocking manner
                 if trial.should_checkpoint():
                     trial.checkpoint()
-                self._pending[trial.train_remote()] = trial
+                self._running[trial.train_remote()] = trial
         except:
             print("Error processing event:", traceback.format_exc())
             if trial.status == Trial.RUNNING:
-                self._return_resources(trial.resources)
-                trial.stop(error=True)
+                self._stop_trial(trial, error=True)
 
     def _get_runnable(self):
+        """
+        TODO(rliaw): At some point this should not be a linear lookup.
+        """
```

Comment:
This should be fine, linear scans are ok for up to thousands of items at our control loop granularity.

Reply:
I think 10k items is easily in scope (especially with when you do early termination and keep all trials in the queue)